File: samples/IterativeSegmentTree.pas
program IterativeSegmentTree;

const
  MAXN = 100000; // Maximum size of the array

var
  N: integer; // Actual size of the array
  tree: array[0..2 * MAXN - 1] of longint;

// Builds the segment tree from an initial array (optional)
procedure Build(arr: array of longint);
var
  i: integer;
begin
  for i := 0 to N - 1 do
    tree[N + i] := arr[i]; // Initialize leaves
  for i := N - 1 downto 1 do
    tree[i] := tree[2 * i] + tree[2 * i + 1]; // Build parents
end;

// Point increment update
procedure Update(idx: integer; val: longint);
begin
  idx := idx + N; // Adjust to tree index
  tree[idx] := tree[idx] + val; // Increment leaf
  while idx > 1 do
  begin
    idx := idx div 2; // Move to parent
    tree[idx] := tree[2 * idx] + tree[2 * idx + 1]; // Update parent
  end;
end;

// Range sum query
function Query(L, R: integer): longint;
var
  sum: longint;
begin
  sum := 0;
  L := L + N; // Adjust to tree index
  R := R + N; // Adjust to tree index
  while L <= R do
  begin
    if (L mod 2 = 1) then // If L is a right child
    begin
      sum := sum + tree[L];
      inc(L);
    end;
    if (R mod 2 = 0) then // If R is a left child
    begin
      sum := sum + tree[R];
      dec(R);
    end;
    L := L div 2; // Move to parent
    R := R div 2; // Move to parent
  end;
  Query := sum;
end;

begin
  // Example Usage:
  N := 5;
  // Initialize with some values (e.g., from an array)
  // For demonstration, let's manually set initial values in the tree leaves
  tree[N+0] := 1;
  tree[N+1] := 2;
  tree[N+2] := 3;
  tree[N+3] := 4;
  tree[N+4] := 5;
  Build([]); // Call Build if you have an initial array

  // Build the parent nodes for the initial values
  for N - 1 downto 1 do
    tree[i] := tree[2 * i] + tree[2 * i + 1];

  writeln('Initial sum of range [0, 4]: ', Query(0, 4)); // Expected: 15

  Update(2, 10); // Increment element at index 2 by 10
  writeln('Sum of range [0, 4] after update: ', Query(0, 4)); // Expected: 25

  writeln('Sum of range [1, 3]: ', Query(1, 3)); // Expected: 2+13+4 = 19
end.

File: samples/IterativeSegmentTree.py
class SegmentTree:
    def __init__(self, arr):
        self.n = len(arr)
        # Tree size is typically 2 * 2^k where 2^k >= n, rounded up to nearest power of 2
        # For simplicity, we can use 2 * n, but 4 * n is a safer general size for recursive implementations.
        # For iterative, 2*n is often sufficient if we carefully manage indices.
        self.tree = [0] * (2 * self.n) # Initialize with zeros

        # Build the segment tree
        self._build(arr)

    def _build(self, arr):
        # Copy original array elements to the leaves of the tree
        for i in range(self.n):
            self.tree[self.n + i] = arr[i]
        # Build up the tree by summing children
        for i in range(self.n - 1, 0, -1):
            self.tree[i] = self.tree[2 * i] + self.tree[2 * i + 1]

    def update(self, index, delta):
        """
        Increments the value at 'index' by 'delta'.
        """
        # Go to the leaf node corresponding to 'index'
        index += self.n
        self.tree[index] += delta

        # Propagate the update up to the root
        while index > 1:
            index //= 2 # Move to parent
            self.tree[index] = self.tree[2 * index] + self.tree[2 * index + 1]

    def query(self, left, right):
        """
        Queries the sum of elements in the range [left, right) (exclusive right).
        """
        res = 0
        left += self.n
        right += self.n

        while left < right:
            if left % 2 == 1: # If left child, include its value and move to next sibling
                res += self.tree[left]
                left += 1
            if right % 2 == 1: # If right child, include its value and move to previous sibling
                right -= 1
                res += self.tree[right]
            left //= 2 # Move up to parent
            right //= 2 # Move up to parent
        return res

